lib/src/models/base_address.dart

```diff
@@ -0,0 +1,24 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:ledger_cardano/src/utils/constants.dart';
+
+part 'base_address.freezed.dart';
+
+@freezed
+sealed class BaseAddress with _$BaseAddress {
+  BaseAddress._();
+
+  factory BaseAddress.basePaymentKeyStakeKey() = BasePaymentKeyStakeKey;
+
+  factory BaseAddress.basePaymentScriptStakeKey() = BasePaymentScriptStakeKey;
+
+  factory BaseAddress.basePaymentKeyStakeScript() = BasePaymentKeyStakeScript;
+
+  factory BaseAddress.basePaymentScriptStakeScript() = BasePaymentScriptStakeScript;
+
+  late final AddressType baseAddressType = switch (this) {
+    BasePaymentKeyStakeKey() => AddressType.basePaymentKeyStakeKey,
+    BasePaymentScriptStakeKey() => AddressType.basePaymentScriptStakeKey,
+    BasePaymentKeyStakeScript() => AddressType.basePaymentKeyStakeScript,
+    BasePaymentScriptStakeScript() => AddressType.basePaymentScriptStakeScript,
+  };
+}
```

Comment:
For reference, that's an example where constructor/factory cannot be const because you defined `baseAddressType` as an extra field.